xbmc/cores/VideoPlayer/DVDCodecs/DVDFactoryCodec.cpp

```diff
@@ -68,27 +69,27 @@ CDVDVideoCodec* CDVDFactoryCodec::CreateVideoCodec(CDVDStreamInfo &hint, CProces
   {
     for (auto &codec : m_hwVideoCodecs)
     {
-      pCodec.reset(CreateVideoCodecHW(codec.first, processInfo));
+      pCodec = CreateVideoCodecHW(codec.first, processInfo);
       if (pCodec && pCodec->Open(hint, options))
       {
-        return pCodec.release();
+        return pCodec;
       }
     }
     if (!(hint.codecOptions & CODEC_ALLOW_FALLBACK))
       return nullptr;
   }
 
-  pCodec.reset(new CDVDVideoCodecFFmpeg(processInfo));
+  pCodec = std::make_unique<CDVDVideoCodecFFmpeg>(processInfo);
   if (pCodec->Open(hint, options))
   {
-    return pCodec.release();
+    return pCodec;
   }
 
   return nullptr;
```

Comment:
You could properly cast this to the constructor:

```c++
return std::unique_ptr<CDVDVideoCodec>();
```

but...

If you have multiple return objects, we can't take advantage of return value optimization. See if you can follow the steps I used to transform this function for RVO: https://github.com/garbear/xbmc/commit/rvo

Reply:
The only issue there is that the places that call `CDVDFactoryCodec::CreateVideoCodec` rely on checking for `nullptr`, also `CDVDVideoCodec` has pure virtual methods so I don't think it can be default constructed.